### PR TITLE
API: Add entrypoint for plotting

### PR DESCRIPTION
**Pull request by TomAugspurger**
Libraries, including pandas, register backends via entrypoints.

xref #26747 

cc @datapythonista @jakevdp @philippjfr

### CODE REVIEW

File: pandas/plotting/_core.py [hunk 1]
```diff
@@ -1533,7 +1533,10 @@ def hexbin(self, x, y, C=None, reduce_C_function=None, gridsize=None, **kwargs):
         return self(kind="hexbin", x=x, y=y, C=C, **kwargs)
 
 
-def _get_plot_backend(backend=None):
+_backends = {}
+
+
+def _get_plot_backend(backend="matplotlib"):
```
Comment (datapythonista) [inline pandas/plotting/_core.py]: Not sure if I'm missing something, but are we ignoring the `plotting.backend` option now?

Also, I find the implementation a bit complex, if it's possible I'd have a function `_load_backends_from_entry_points` that is called `if not _backends` and the rest of `_get_plot_backend` should simply try the lookup and do the import.

File: pandas/tests/plotting/test_backend.py [hunk 2]
```diff
@@ -36,3 +42,44 @@ def test_backend_is_correct(monkeypatch):
...
+    assert result is mod
+
+
+def test_no_matplotlib_ok():
+    try:
+        import matplotlib  # noqa
+    except ImportError:
+        pass
+    else:
+        raise pytest.skip("matplotlib installed.")
```
Comment (datapythonista) [inline pandas/tests/plotting/test_backend.py]: Don't we have a decorator `skip_if_mpl`? May be I'm missing something, but seems like should be the same.

**Review by datapythonista — CHANGES_REQUESTED** [review]
Couple of comments.

File: pandas/plotting/_core.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/plotting/_core.py]: > Not sure if I'm missing something, but are we ignoring the plotting.backend option now?

Uhmm whoops.

**Comment by jreback** [comment]
this seems like a very complex solution 

why is not the simpler

plotting.backend = ‘altair.pandas_plot’ 
plotting.backend = ‘pandas._matplotlib.plot’

not much more straightforward here?

eg this is a standard import as passed to importlib

**Comment by TomAugspurger** [comment]
We want to decouple the backend name (`altair`) from the implementation. Altair doesn't want and shouldn't need a top-level `altair.pandas_plot` just to implement this backend (which may even be implemented in a separate package).

**Comment by jreback** [comment]
and how does my suggestion not do this is? it is fully decoupled

**Comment by TomAugspurger** [comment]
I guess I don't understand your suggestion then. You say 

```
plotting.backend = ‘altair.pandas_plot’
```

which seems to require a top-level `altair.pandas_plot`?

**Comment by jreback** [comment]
not at all
these are examples of what you could pass

**Comment by jakevdp** [comment]
The reason I proposed this is because I believe it's cleaner for the implementation's code path to be separate from how the user specifies it. My goal would be for a user to write
```
plotting.backend = 'altair'
```
and for this to cause the plotting backend to use Altair.

Before this solution, that would require me to pollute the Altair package's top-level namespace with APIs that are irrelevant to users of the package. I think entrypoints is a clean solution to that problem (decoupling of specification from code paths), and one that is in common use among packages in the jupyter ecosystem.

**Comment by jakevdp** [comment]
I just want to avoid a situation where users have to remember, "for Matplotlib use ``plotting.backend = 'matplotlib'``", "for Bokeh, use ``plotting.backend = 'bokeh. _pandas_api``", for Altair use ``plotting.backend = 'altair_pandas_backend'``"...

Seems like it would lead to more confusion than necessary.

**Comment by TomAugspurger** [comment]
Thanks @jakevdp. Agreed entirely.

**Comment by jreback** [comment]
ok I see your point @jakevdp and I agree that is nice to de-couple the user supplied key to plotting.backend, but then again it *does* couple the api of the package to pandas directly, meaning we then *need* to make an update in pandas to actually use a new package / api. 

So this is nice, I would then *also* allow a fully qualified module.function call to succeed

**Comment by TomAugspurger** [comment]
> it does couple the api of the package to pandas directly, meaning we then need to make an update in pandas to actually use a new package / api.

I may be misunderstanding, but I don't see how the change to use entrypoints affects that. If you're saying that we'll need to update pandas so that `set_option('plotting.backend', 'altair')` works, then that's not correct. The only change required after this is that altair (or pdvega, whatever package) adds an EntryPoint for `pandas_plotting_backends` in its setup.py.

Basically, pandas adds an EntryPoint "group" in our setup.py. And third party libraries add items to that group in their setup.pys. Pandas and the library don't need to talk to each other directly to register a backend. We talk through `pkg_resources`.

> So this is nice, I would then also allow a fully qualified module.function call to succeed

We do allow that (see `test_register_import`).

**Comment by jreback** [comment]
> Basically, pandas adds an EntryPoint "group" in our setup.py. And third party libraries add items to that group in their setup.pys. Pandas and the library don't need to talk to each other directly to register a backend. We talk through `pkg_resources`.
> 

ok so if this works then that's great. +1.

**Comment by TomAugspurger** [comment]
Yep, this should work out just fine.

**Comment by jreback** [comment]
@datapythonista ok on this?

File: pandas/plotting/_core.py [hunk 3]
```diff
@@ -1533,6 +1533,53 @@ def hexbin(self, x, y, C=None, reduce_C_function=None, gridsize=None, **kwargs):
...
+
+    Parameters
+    ----------
+    backend : str
+        The identifier for the backend. Either an entrypoint item registered
+        with pkg_resources, or a module name.
+
+    Notes
+    -----
+    Modifies _backcends with imported backends as a side effect.
```
Comment (datapythonista) [inline pandas/plotting/_core.py]:
```suggestion
    Modifies _backends with imported backends as a side effect.
```

File: pandas/plotting/_core.py [hunk 4]
```diff
@@ -1533,6 +1533,53 @@ def hexbin(self, x, y, C=None, reduce_C_function=None, gridsize=None, **kwargs):
...
+
+    Returns
+    -------
+    backend : types.ModuleType
```
Comment (datapythonista) [inline pandas/plotting/_core.py]:
```suggestion
    types.ModuleType
```

We've been just naming the output when a function returns a tuple. In this case seems reasonable, but we had many "transform -> transformed" names that didn't add much value.

**Review by datapythonista — APPROVED** [review]
lgtm, just couple of small things in the docstring.

**Comment by TomAugspurger** [comment]
Thanks, fixed.

**Comment by jreback** [comment]
thanks @TomAugspurger

File: pandas/tests/plotting/test_backend.py [hunk 5]
```diff
@@ -36,3 +42,44 @@ def test_backend_is_correct(monkeypatch):
         pandas.set_option("plotting.backend", "matplotlib")
     except ImportError:
         pass
+
+
+@td.skip_if_no_mpl
+def test_register_entrypoint():
```
Comment (jbrockmendel) [inline pandas/tests/plotting/test_backend.py]: this is failing for me locally

Comment (TomAugspurger) [reply pandas/tests/plotting/test_backend.py]: What are you seeing?

Comment (TomAugspurger) [reply pandas/tests/plotting/test_backend.py]: Oh, I'm guessing it's a `KeyError: 'pandas_plotting_backends'`?

You'll need to re-run `python -m pip install -e .` in your pandas directory. This adds the entrypoint.

Comment (jbrockmendel) [reply pandas/tests/plotting/test_backend.py]: Yep that does fix it, thanks

**Pull request merged**